include/cb_loop.h

```diff
@@ -0,0 +1,110 @@
+/**
+ * @brief RL Inference API definition.
+ *
+ * @file cb_loop.h
+ * @author Rajan Chari et al
+ * @date 2018-07-18
+ */
+#pragma once
+#include "action_flags.h"
+#include "base_loop.h"
+#include "err_constants.h"
+#include "factory_resolver.h"
+#include "future_compat.h"
+#include "ranking_response.h"
+#include "rl_string_view.h"
+#include "sender.h"
+
+#include <functional>
+#include <memory>
+
+namespace reinforcement_learning
+{
+class cb_loop : public base_loop
+{
+public:
+  using base_loop::base_loop;
+
+  /**
+   * @brief Choose an action, given a list of actions, action features and context features. The
+   * inference library chooses an action by creating a probability distribution over the actions
+   * and then sampling from it.
+   * @param event_id  The unique identifier for this interaction.  The same event_id should be used when
+   *                  reporting the outcome for this action.
+   * @param context_json Contains action, action features and context features in json format
+   * @param resp Ranking response contains the chosen action, probability distribution used for sampling actions and
+   * ranked actions
+   * @param status  Optional field with detailed string description if there is an error
+   * @return int Return error code.  This will also be returned in the api_status object
+   */
+  int choose_rank(const char* event_id, string_view context_json, ranking_response& resp, api_status* status = nullptr);
```

Comment:
I agree with the general reasoning that string types should be passed in with a size to prevent certain failure modes.  For example, certain languages, don't have null terminated strings.  

However, string_view class in our public interface introduces a type with very large surface area.  I'd like to see this replaced with a simple "buffer" like class with a small surface area.  for example buffer or string_buffer struct with a const char* and a size.

Reply:
I have 2 primary issues with this:
1) We are introducing a type in our own public interface that we have little control over.
2) A large unnecessary surface area makes usage automated binding tools like SWIG difficult or impossible.

Reply:
So the requirement for this API change is to provide a C style API aiming mostly at bindings, and not a C++ API, correct?

Reply:
Some clarification:

With (1) I am primarily concerned with exposing classes with implementation and contracts that we don't have control over and unanticipated consequences of doing that.

Issues in this category are things like someone linking with a precompiled version of our libraries.  For example, someone building in Debug mode and linking with our libraries in Release mode or vice versa.  Classes allocated in client code but passed into our code can have undefined behavior.  Classes defined in the interface may throw under conditions that we are unaware of etc.  So C++ is fine as long as it's our classes.  A buffer classes that holds char * and size is perfectly acceptable and even preferable.  This keeps the spirit in with string_view was introduced intact.  i.e. we should carry the buffer and its length together as a unit.

With (2) The concern is binding with automated binding tools.  I am hoping that we get a number of languages for free with this approach... for example.. javascript bindings would be nice.

For example, neither of these 2 concerns preclude the use of a C++ iterator type interface.  We do implement this somewhere.

Reply:
@rajan-chari Please see my comment: https://github.com/VowpalWabbit/reinforcement_learning/pull/572/files#r1153409981

I think there are two issues being combined here and I don't see that value of combining them. Yes, having a ABI stable C interface is useful. But it is also useful to have a C++ interface. Why must they be the same thing? It is trivial to implement the ABI stable C interface on top of the core library being written in C++.

Users then get the choice of C if they need ABI stability, or C++ if that works best for them.

Reply:
The objection to my concerns and proposal is not clear.

Is the claim that the right C++ approach is to use string_view and other std types in our C++ interface?